crates/red_knot/src/ast_ids.rs

```diff
@@ -0,0 +1,322 @@
+use std::fmt::Formatter;
+use std::marker::PhantomData;
+
+use rustc_hash::FxHashMap;
+
+use ruff_index::{Idx, IndexVec};
+use ruff_python_ast::visitor::preorder;
+use ruff_python_ast::visitor::preorder::{PreorderVisitor, TraversalSignal};
+use ruff_python_ast::{
+    AnyNodeRef, AstNode, ModModule, NodeKind, Parameter, Stmt, StmtAnnAssign, StmtAssign,
+    StmtAugAssign, StmtClassDef, StmtFunctionDef, StmtImport, StmtImportFrom, StmtTypeAlias,
+    TypeParam, TypeParamParamSpec, TypeParamTypeVar, TypeParamTypeVarTuple,
+};
+use ruff_text_size::{Ranged, TextRange};
+
+#[ruff_index::newtype_index]
+pub struct AstId;
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub struct FileAstId<N: HasAstId> {
+    ast_id: AstId,
+    _marker: PhantomData<fn() -> N>,
+}
+
+impl<N: HasAstId> FileAstId<N> {
+    pub fn upcast<M: HasAstId>(self) -> FileAstId<M>
+    where
+        N: Into<M>,
+    {
+        FileAstId {
+            ast_id: self.ast_id,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<N: HasAstId> Copy for FileAstId<N> {}
+impl<N: HasAstId> Clone for FileAstId<N> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+pub struct AstIds {
+    ids: IndexVec<AstId, SyntaxNodeKey>,
+    reverse: FxHashMap<SyntaxNodeKey, AstId>,
+}
+
+impl AstIds {
+    // TODO rust analyzer doesn't allocate an ID for every node. It only allocates ids for
+    //  nodes with a corresponding HIR element, that is nodes that are definitions.
+    pub fn from_module(module: &ModModule) -> Self {
+        let mut visitor = AstIdsVisitor::default();
+
+        // TODO: visit_module?
+        // Make sure we visit the root
+        visitor.create_id(module);
+        visitor.visit_body(&module.body);
+
+        while let Some(deferred) = visitor.deferred.pop() {
+            match deferred {
+                DeferredNode::FunctionDefinition(def) => {
+                    def.visit_preorder(&mut visitor);
+                }
+                DeferredNode::ClassDefinition(def) => def.visit_preorder(&mut visitor),
```

Comment:
I think that functions should be deferred but class bodies probably should not be, because at runtime the class body is evaluated immediately to create the class, and anything referenced in it needs to already be defined. and we need the class body to know the members of the class

Reply:
Maybe, I don't know. I think the ID generation is tightly tied to the HIR design and it's currently unclear to me what the HIR should capture exactly (as discussed over Discord). 

Overall, the goal of the HIR is to isolated changes and I could see it being beneficial if adding a class member doesn't change the IDs of functions coming after. The assignment of the ID also doesn't mean that these nodes don't have IDs. It only means that they have less stable IDs than the class itself.